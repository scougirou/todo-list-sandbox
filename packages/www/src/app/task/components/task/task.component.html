<div>
  <span>
    {{ task.title }}
  </span>
  <span>
    <mat-icon >delete</mat-icon>
  </span>
</div>
