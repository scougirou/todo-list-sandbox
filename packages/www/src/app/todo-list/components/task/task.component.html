<div>
  <span>
    <mat-checkbox
      [disabled]="isTaskDone()"
      [checked]="isTaskDone()"
      (change)="markTaskAsDone()">
    </mat-checkbox>
  </span>
  <span class="task-title" [class.complete]="isTaskDone()">
    {{ task.title }}
  </span>
  <span>
    <mat-icon class="delete" (click)="deleteTask()">delete</mat-icon>
  </span>
</div>
