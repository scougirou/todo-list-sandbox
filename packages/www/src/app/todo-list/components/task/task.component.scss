

.complete {
  color: gray;
  text-decoration: line-through;
}

.delete {
  cursor: pointer;
}
