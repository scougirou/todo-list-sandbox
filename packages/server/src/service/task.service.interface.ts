import { TaskDto } from '../model/task.dto';
import { TaskStatusEnum } from '../model/task-status.enum';

export interface TasksServiceInterface {
  getTaskList(): TaskDto[];

  addTask(task: TaskDto): void;

  deleteTask(taskId: string): string | undefined;

  completeTask(taskId: string): TaskDto;

  filterBy(filterStatus: TaskStatusEnum): TaskDto[];
}
