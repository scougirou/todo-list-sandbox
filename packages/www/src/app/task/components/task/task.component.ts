import { Component, Input, OnInit } from '@angular/core';
import { TaskDto } from '../../model/task.dto';

@Component({
  selector: 'task-row',
  templateUrl: './task.component.html',
  styleUrls: ['./task.component.scss']
})
export class TaskComponent implements OnInit {
  @Input()
  task: TaskDto = {};

  constructor() { }

  ngOnInit(): void {
  }

}
