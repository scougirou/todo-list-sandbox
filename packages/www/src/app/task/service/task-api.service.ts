import { Injectable } from '@angular/core';
import { TaskDto } from '../model/task.dto';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../../environments/environment';
import { BehaviorSubject } from 'rxjs';

const GET_TASK_LIST_URI = '/tasks'
const ADD_TASK_LIST_URI = '/tasks'

@Injectable({
  providedIn: 'root'
})
export class TaskApiService {
  taskListSubject: BehaviorSubject<TaskDto[]>;

  constructor(private readonly httpClient: HttpClient) {
    this.taskListSubject = new BehaviorSubject<TaskDto[]>([]);
  }

  async addTask(task: TaskDto): Promise<void> {
    await this.httpClient.post(`${environment.apiUrl}${ADD_TASK_LIST_URI}`, task ).toPromise();
    this.updateTaskList();
  }

  private async updateTaskList(): Promise<void> {
    const taskList = await this.getAllTasks();
    this.taskListSubject.next(taskList);
  }

  private async getAllTasks(): Promise<TaskDto[]> {
    const taskList = await this.httpClient.get<TaskDto[]>(`${environment.apiUrl}${GET_TASK_LIST_URI}`).toPromise();
    return taskList;
  }
}
