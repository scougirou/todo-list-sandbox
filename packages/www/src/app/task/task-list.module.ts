import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TaskListPageComponent } from './list-page/task-list-page.component';
import { TaskComponent } from './components/task/task.component';
import { HttpClientModule } from '@angular/common/http';
import { AddTaskComponent } from './components/add-task/add-task.component';
import { ReactiveFormsModule } from '@angular/forms';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';

@NgModule({
  declarations: [
    TaskListPageComponent,
    TaskComponent,
    AddTaskComponent
  ],
  imports: [
    CommonModule,
    HttpClientModule,
    ReactiveFormsModule,
    MatIconModule,
    MatButtonModule,
  ],
})
export class TaskListModule { }
