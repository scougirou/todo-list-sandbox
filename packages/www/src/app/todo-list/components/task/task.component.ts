import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { TaskDto } from '../../model/task.dto';
import { TaskStatusEnum } from '../../model/task-status.enum';

@Component({
  selector: 'task-row',
  templateUrl: './task.component.html',
  styleUrls: ['./task.component.scss']
})
export class TaskComponent implements OnInit {
  @Input()
  task: TaskDto = {};

  @Output()
  markTaskAsDoneOutput = new EventEmitter<string>();
  @Output()
  deleteTaskOutput = new EventEmitter<string>();

  constructor() { }

  ngOnInit(): void {
  }

  isTaskDone(): boolean {
    return this.task.status === TaskStatusEnum.DONE;
  }

  markTaskAsDone() {
    this.markTaskAsDoneOutput.emit(this.task.id);
  }

  deleteTask() {
    this.deleteTaskOutput.emit(this.task.id)
  }
}
