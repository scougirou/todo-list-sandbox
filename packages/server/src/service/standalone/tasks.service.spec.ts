import { Test, TestingModule } from '@nestjs/testing';
import { TasksService } from './tasks.service';
import { TaskDto } from '../../model/task.dto';
import { TaskStatusEnum } from '../../model/task-status.enum';

describe('TasksService', () => {
  let service: TasksService;

  let dummyTask: TaskDto;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [TasksService],
    }).compile();

    service = module.get<TasksService>(TasksService);

    dummyTask = {
      id: '12345',
      createdAt: new Date(1635166915096),
      modifiedAt: new Date(1635166915096),
      title: 'Dummy todo-list',
      status: TaskStatusEnum.TODO,
    };
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should be able to add new tasks', () => {
    service.addTask(dummyTask);

    expect(service.taskList).toMatchSnapshot();
  });

  it('should be able to get tasks', () => {
    service.taskList = new Map<string, TaskDto>();
    service.taskList.set('12345', dummyTask);
    service.taskList.set('67890', dummyTask);

    expect(service.getTaskList()).toStrictEqual([dummyTask, dummyTask]);
  });

  it('should update todo-list status to done', () => {
    service.taskList = new Map<string, TaskDto>();
    service.taskList.set('12345', dummyTask);

    service.completeTask('12345');

    expect(service.taskList.get('12345').status).toStrictEqual(
      TaskStatusEnum.DONE,
    );
  });

  it('should delete task', () => {
    service.taskList = new Map<string, TaskDto>();
    service.taskList.set('12345', dummyTask);

    service.deleteTask('12345');

    expect(service.taskList).toStrictEqual(new Map<string, TaskDto>());
  });

  it('should filter tasks', () => {
    service.taskList = new Map<string, TaskDto>();
    // Avoid mutating a common object during tests, mitigate side effects
    const dummyTask3 = Object.assign({}, dummyTask);
    const dummyTask4 = Object.assign({}, dummyTask);
    dummyTask3.status = TaskStatusEnum.DONE;
    dummyTask3.id = 'dummyTask3';
    dummyTask4.status = TaskStatusEnum.DONE;
    dummyTask4.id = 'dummyTask4';
    service.taskList.set(dummyTask.id, dummyTask);
    service.taskList.set(dummyTask.id, dummyTask);
    service.taskList.set(dummyTask3.id, dummyTask3);
    service.taskList.set(dummyTask4.id, dummyTask4);

    const result = service.filterBy(TaskStatusEnum.DONE);

    expect(result).toStrictEqual([dummyTask3, dummyTask4]);
  });
});
