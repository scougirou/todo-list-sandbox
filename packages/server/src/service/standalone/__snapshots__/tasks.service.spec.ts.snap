// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksService should be able to add new tasks 1`] = `
Map {
  "12345" => Object {
    "createdAt": 2021-10-25T13:01:55.096Z,
    "id": "12345",
    "modifiedAt": 2021-10-25T13:01:55.096Z,
    "status": "TODO",
    "title": "Dummy todo-list",
  },
}
`;
